### ISSUE TITLE
check_all dynamic typing issues

### ISSUE THREAD

**Issue by skrukwa** [pyta-uoft/pyta#882]
Hello,

I have noticed that things like `for` loops and `while` loops that change the types of variables can cause pyta to report incorrect warnings.

An example is as follows.
```python
def f() -> int:
    """while loop test function"""
    var = None
    while True:
        assert var is None or isinstance(var, tuple)

        if var is None:
            var = 1, 2
        else:
            return var[0]  # pyta reports E1136 (unsubscriptable-object) warning


if __name__ == '__main__':
    import python_ta
    python_ta.check_all()
```

I assume this is probably not an easy fix, and is probably already a known issue, but thought I would post it here formally. Feel free to do with thread as you see fit.

**Comment by skrukwa** [comment]
On a side note, another (easier to fix) issue results from pyta not treating `raise` like `return`.

An example is as follows.

```python
def g(a: int) -> int:
    """if statements test function"""
    if a % 3 == 0:
        result = 3
    elif a % 2 == 0:
        result = 2
    else:
        raise NotImplementedError

    return result  # pyta reports E9969 (possibly-undefined) warning
```

**Comment by david-yz-liu** [comment]
Thanks for reporting these issues. The first issue no longer is triggered (fixed by pylint) and the second has been fixed in #884.
